src/iop/filmicrgb.c

```diff
@@ -1502,6 +1504,363 @@ static inline void filmic_chroma_v2_v3(const float *const restrict in, float *co
+  float chroma_final = Ych_final[1] * Ych_final[0];                 // c1
+
+  // fit a linear model `chroma = f(y)`:
+  // `chroma = c1 + (yc - y1) * (c2 - c1) / (y2 - y1)`
+  // where `(yc - y1)` is user-defined as `saturation * (y2 - y1)`
+  // so `chroma = c1 + saturation * (c2 - c1)`
+  // when saturation = 0, we stay at the saturation-invariant final chroma
+  // when saturation > 1, we go back towards the initial chroma before tone-mapping
+  // when saturation < 0, we amplify the initial -> final chroma change
+  const float delta_chroma = saturation * (chroma_original - chroma_final);
+
+  const int filmic_brightens = (Ych_final[0] > Ych_original[0]);
+  const int filmic_resat = (chroma_original < chroma_final);
+  const int filmic_desat = (chroma_original > chroma_final);
+  const int user_resat = (saturation > 0.f);
+  const int user_desat = (saturation < 0.f);
+
+  chroma_final = (filmic_brightens && filmic_resat)
+                      ? chroma_original // force original lower sat if brightening
+                  : ((user_resat && filmic_desat) || user_desat)
+                      ? chroma_final + delta_chroma // allow resaturation only if filmic desaturated, allow desat anytime
+                      : chroma_final;
+
+  Ych_final[1] = fmaxf(chroma_final / Ych_final[0], 0.f);
+}
+
+// Pipeline and ICC luminance is CIE Y 1931
+// Kirk Ych/Yrg uses CIE Y 2006
+// 1 CIE Y 1931 = 0.945302269001 CIE Y 2006, so we need to adjust that
+#define CIE_Y_1931_to_CIE_Y_2006(x) 0.945302269001f * x
+
+
+static inline void gamut_check_RGB(dt_aligned_pixel_t RGB_in, dt_aligned_pixel_t Ych_in,
+                                   const dt_colormatrix_t matrix_in, const dt_colormatrix_t matrix_out,
+                                   const float display_black, const float display_white)
+{
+  const int max_iter = 8;
+
+  int iter = 0;
+  float max_pix = fmaxf(fmaxf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+  while(max_pix > display_white && iter < max_iter)
+  {
+    // Perform exposure correction to fit back in gamut
+    // This gives us the closest in-gamut pixel at same saturation.
+    dt_aligned_pixel_t RGB_darkened = { 0.f };
+    for_each_channel(c, aligned(RGB_darkened)) RGB_darkened[c] = RGB_in[c] / max_pix;
+    dt_aligned_pixel_t Ych_darkened = { 0.f };
+    pipe_RGB_to_Ych(RGB_darkened, matrix_in, Ych_darkened);
+
+    // Remapping goal:
+    // 1. at constant hue and luminance, march toward the chroma of the closest same-saturation in-gamut color.
+    //    Since same-saturation is darker, its chroma should be much lower than the current.
+    // 2. iterate until convergence.
+    Ych_in[1] = Ych_darkened[1] * Ych_darkened[0] / Ych_in[0];
+    Ych_to_pipe_RGB(Ych_in, matrix_out, RGB_in);
+    max_pix = fmaxf(fmaxf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+    iter++;
+  }
+
+  // Finally, look for negatives.
+  float min_pix = fminf(fminf(RGB_in[0], RGB_in[1]), RGB_in[2]);
+  iter = 0;
+  while(min_pix < display_black && iter < max_iter)
+  {
```

Comment:
I see you chose to update both luminance and chroma here. This will lift up the shadows a bit. Is there a particular reason for this choice (aside of the heuristic using the brightened and clamped RGB here in iteration)? If there was a closed-form or numerical solution for finding the largest chroma for the color that is in-gamut, should such a solution be used here instead? I can try in SymPy (or pen and paper) if such a solution could be found.